tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/SymbStateRewriterImplWithArrays.scala

```diff
@@ -9,6 +9,16 @@ import at.forsyte.apalache.tla.lir.convenience.tla
 import at.forsyte.apalache.tla.lir.oper._
 import at.forsyte.apalache.tla.lir.UntypedPredefs._
 
+/**
+ * This class extends SymbStateRewriterImpl with encoding rules for the encoding with SMT arrays.
+ * TODO: enforce that _solverContext.config.smtEncoding is correct?
```

Comment:
Just add `require(_solverContext.config.smtEncoding == ...)`?

Reply:
I initially though of enforcing this at every point the rewriter is constructed, but that is not nice. Adding `require(...)` to the base rewriter and `override require(...)` to the rewriter with arrays does not work, nor does adding `require(cond)` and overriding `cond`. Is there a clean solution to this?

Reply:
I thought `require` was just a part of the constructor. Not sure that you can override it.

Reply:
Exactly. If we want to put `require(cond)` in the parent constructor, we need to make sure `cond` is overwritten before it is accessed. I did a search and it seems that there is no way to do any computation in the child class before the parent constructor is called, so we end up requiring `_solverContext.config.smtEncoding == "oopsla19"` for the arrays encoding.

I could either put the `require` only on the child rewriter or precede the construction of every rewriter with the `require`. Any thoughts?

Reply:
Will be done in a follow-up PR.